libopae/plugins/xfpga/enum.c

```diff
@@ -477,134 +478,65 @@ STATIC fpga_result enum_afu(const char *sysfspath, const char *name,
-	if (stat(sysfspath, &stats) != 0) {
-		FPGA_MSG("stat failed: %s", strerror(errno));
-		return FPGA_NO_DRIVER;
+	if (region_count <= 0) {
+		FPGA_MSG("Not found fpga regions");
+		return FPGA_NO_DRIVER;;
 	}
 
-	if (!S_ISDIR(stats.st_mode))
-		return FPGA_OK;
-
-	res = readlink(sysfspath, spath, sizeof(spath));
-	if (-1 == res) {
-		FPGA_MSG("Can't read link");
-		return FPGA_NO_DRIVER;
-	}
+	for (i = 0; i < region_count; i++) {
 
-	pdev = add_dev(sysfspath, "", list);
-	if (!pdev) {
-		FPGA_MSG("Failed to allocate device");
-		return FPGA_NO_MEMORY;
-	}
+		region  = (sysfs_fpga_region *)sysfs_get_region(i);
 
-	// Find the BDF from the link path.
-	spath[res] = 0;
-	p = strrchr(spath, '/');
-	if (!p) {
-		FPGA_MSG("Invalid link");
-		return FPGA_NO_DRIVER;
-	}
-	*p = 0;
-	p = strrchr(spath, '/');
-	if (!p) {
-		FPGA_MSG("Invalid link");
-		return FPGA_NO_DRIVER;
-	}
-	*p = 0;
-	p = strrchr(spath, '/');
-	if (!p) {
-		FPGA_MSG("Invalid link");
-		return FPGA_NO_DRIVER;
-	}
-	++p;
-
-	//           11
-	// 012345678901
-	// ssss:bb:dd.f
-	f = 0;
-	sscanf_s_i(p + 11, "%d", &f);
-
-	pdev->function = (uint8_t)f;
-	*(p + 10) = 0;
-
-	d = 0;
-	sscanf_s_u(p + 8, "%x", &d);
-
-	pdev->device = (uint8_t)d;
-	*(p + 7) = 0;
-
-	b = 0;
-	sscanf_s_u(p + 5, "%x", &b);
-
-	pdev->bus = (uint8_t)b;
-	*(p + 4) = 0;
-
-	s = 0;
-	sscanf_s_u(p, "%x", &s);
-	pdev->segment = (uint16_t)s;
-
-	// read the vendor and device ID from the 'device' path
-	uint32_t x = 0;
-	char vendorpath[SYSFS_PATH_MAX];
-	snprintf_s_s(vendorpath, SYSFS_PATH_MAX, "%s/device/vendor", sysfspath);
-	result = sysfs_read_u32(vendorpath, &x);
-	if (result != FPGA_OK)
-		return result;
-	pdev->vendor_id = (uint16_t)x;
-
-	char devicepath[SYSFS_PATH_MAX];
-	snprintf_s_s(devicepath, SYSFS_PATH_MAX, "%s/device/device", sysfspath);
-	result = sysfs_read_u32(devicepath, &x);
-	if (result != FPGA_OK)
-		return result;
-	pdev->device_id = (uint16_t)x;
-
-	// Find the FME and AFU devices.
-	dir = opendir(sysfspath);
-	if (NULL == dir) {
-		FPGA_MSG("Can't open directory: %s", sysfspath);
-		return FPGA_NO_DRIVER;
-	}
+		if (!region) {
+			FPGA_MSG("failed to enum regions");
+			return FPGA_NO_DRIVER;
+		}
 
-	while ((dirent = readdir(dir)) != NULL) {
-		if (!strcmp(dirent->d_name, "."))
-			continue;
-		if (!strcmp(dirent->d_name, ".."))
-			continue;
+		pdev = add_dev(region->region_path, "", list);
+		if (!pdev) {
+			FPGA_MSG("Failed to allocate device");
+			return FPGA_NO_MEMORY;
+		}
+		// Assign bus, function, device
+		// segment,device_id ,vendor_id
+		pdev->function       = region->function;
+		pdev->segment        = region->segment;
+		pdev->bus            = region->bus;
+		pdev->device         = region->device;
+		pdev->device_id      = region->device_id;
+		pdev->vendor_id      = region->vendor_id;
+
+		// Enum fme
+		if (region->fme) {
+			result = enum_fme(region->fme->res_path,
+							region->fme->res_name, pdev);
+			if (result != FPGA_OK) {
+				FPGA_ERR("Failed to enum FME");
```

Comment:
Do we want to stop and return an error code here?

Reply:
fixed